packages/api/src/auth.js

```diff
@@ -68,11 +69,40 @@ export function withApiOrMagicToken (handler) {
   }
 }
 
+/**
+ * Middleware: verify the request is authenticated with a valid api token
+ * for a user that has pinning enabled.
+ * On successful login, adds `auth.user` and `auth.auth.token` to the Request
+ *
+ * @param {import('itty-router').RouteHandler} handler
+ * @returns {import('itty-router').RouteHandler}
+ */
+export function withPinningEnabledUser (handler) {
+  /**
+   * @param {Request} request
+   * @param {import('./env').Env}
+   * @returns
+   */
+  return async (request, env, ctx) => {
+    const token = getTokenFromRequest(request, env)
+    const apiToken = await tryWeb3ApiToken(token, env)
+    if (apiToken) {
+      if (apiToken.user.pinningEnabled) {
+        request.auth = { authToken: apiToken, user: apiToken.user }
+        env.sentry && env.sentry.setUser(apiToken.user)
+        return handler(request, env, ctx)
+      }
+      throw new PinningNotEnabledError()
+    }
+    throw new UnrecognisedTokenError()
```

Comment:
I'd just have this middleware take an authenticated request.

```suggestion
    const { user } = request.auth
    if (!user.pinningEnabled) {
      throw new PinningNotEnabledError()
    }
    return handler(request, env, ctx)
```

Although I think it would be preferable to store this information in a separate table that lists user IDs with authorisations. We're having an admin area built and it would be safer if it didn't ever write to the user table.

So instead of checking for `user.pinningEnabled` you'd pass the user ID to some function that would check the user's ID is present in the pinning service authorisations table.

Reply:
Have now implemented. Using a separate table `pinning_authorisation` and a DB method to check for the existence of a row for the authenticated user.